mypy errors on CI

There are some existing mypy errors on CI. Mostly having to do with many types of arrays being allowed, but code expecting a certain type. I am hoping to take a look.

```python-traceback
>>> mypy dask_ml/metrics
dask_ml/metrics/regression.py:102: error: "ndarray" has no attribute "chunks"
dask_ml/metrics/pairwise.py:30: error: "ndarray" has no attribute "to_delayed"
dask_ml/metrics/pairwise.py:37: error: "ndarray" has no attribute "chunks"
dask_ml/metrics/pairwise.py:42: error: "ndarray" has no attribute "chunks"
dask_ml/metrics/pairwise.py:58: error: "ndarray" has no attribute "chunks"
dask_ml/metrics/pairwise.py:59: error: "ndarray" has no attribute "map_blocks"
dask_ml/metrics/pairwise.py:136: error: Argument 2 to "check_pairwise_arrays" has incompatible type "Optional[ndarray]"; expected "ndarray"
dask_ml/metrics/pairwise.py:144: error: Argument 2 to "check_pairwise_arrays" has incompatible type "Optional[ndarray]"; expected "ndarray"
dask_ml/metrics/pairwise.py:161: error: Argument 2 to "check_pairwise_arrays" has incompatible type "Optional[ndarray]"; expected "ndarray"
dask_ml/metrics/pairwise.py:176: error: Argument 2 to "check_pairwise_arrays" has incompatible type "Optional[ndarray]"; expected "ndarray"
dask_ml/metrics/pairwise.py:211: error: Argument 2 to "check_pairwise_arrays" has incompatible type "Optional[ndarray]"; expected "ndarray"
Found 11 errors in 2 files (checked 5 source files)
```
```python-traceback
>>> mypy dask_ml/preprocessing
dask_ml/preprocessing/label.py:200: error: Incompatible default for argument "uniques" (default has type "None", argument has type "ndarray")
dask_ml/preprocessing/label.py:202: error: "ndarray" has no attribute "cat"
dask_ml/preprocessing/label.py:212: error: "ndarray" has no attribute "cat"
dask_ml/preprocessing/data.py:39: error: Incompatible types in assignment (expression has type "float", variable has type "ndarray")
dask_ml/preprocessing/data.py:78: error: Need type annotation for 'n_features_in_'
dask_ml/preprocessing/data.py:145: error: Need type annotation for 'n_features_in_'
dask_ml/preprocessing/data.py:229: error: Argument 1 to "_handle_zeros_in_scale" has incompatible type "List[float]"; expected "ndarray"
dask_ml/preprocessing/data.py:230: error: Need type annotation for 'n_features_in_'
dask_ml/preprocessing/data.py:335: error: Missing return statement
dask_ml/preprocessing/data.py:371: error: "ndarray" has no attribute "map_blocks"
dask_ml/preprocessing/data.py:385: error: "ndarray" has no attribute "map_blocks"
dask_ml/preprocessing/data.py:391: error: "ndarray" has no attribute "map_blocks"
dask_ml/preprocessing/data.py:399: error: "ndarray" has no attribute "map_blocks"
dask_ml/preprocessing/data.py:764: error: Incompatible types in assignment (expression has type "Tuple[Tuple[Any, ...], Any]", variable has type "ndarray")
dask_ml/preprocessing/data.py:774: error: Incompatible types in assignment (expression has type "Tuple[Any, ...]", variable has type "ndarray")
dask_ml/preprocessing/data.py:989: error: Incompatible types in assignment (expression has type "Tuple[Tuple[Any, ...], Any]", variable has type "ndarray")
dask_ml/preprocessing/data.py:999: error: Incompatible types in assignment (expression has type "Tuple[Any, ...]", variable has type "ndarray")
dask_ml/preprocessing/_encoders.py:119: error: Incompatible default for argument "dtype" (default has type "Type[floating[Any]]", argument has type "dtype[Any]")
dask_ml/preprocessing/_block_transformer.py:74: error: Need type annotation for 'func'
Found 19 errors in 4 files (checked 5 source files)

--- a/dask_ml/_typing.py
+++ b/dask_ml/_typing.py
@@ -1,4 +1,4 @@
-from typing import Any, Sequence, Type, TypeVar, Union
+from typing import Any, Sequence, TypeVar, Union
 
 import dask.dataframe as dd
 import numpy as np
@@ -8,7 +8,7 @@ from pandas import DataFrame, Index, Series
 try:
     from numpy.typing import DTypeLike
 except ImportError:
-    DTypeLike = Any
+    DTypeLike = Any  # type: ignore
 
 
 AnyArrayLike = TypeVar("AnyArrayLike", Index, Series, Array, np.ndarray)

--- a/dask_ml/_typing.py
+++ b/dask_ml/_typing.py
@@ -1,9 +1,11 @@
-from typing import TypeVar, Union
+from typing import TypeVar, Union, Sequence
 
 import dask.dataframe as dd
 import numpy as np
 from dask.array import Array
 from pandas import DataFrame, Index, Series
+from numpy.typing import DTypeLike
+
 
 AnyArrayLike = TypeVar("AnyArrayLike", Index, Series, Array, np.ndarray)
 ArrayLike = TypeVar("ArrayLike", Array, np.ndarray)
@@ -12,3 +14,4 @@ SeriesType = Union[dd.Series, Series]
 DataFrameType = Union[DataFrame, dd.DataFrame]
 Number = Union[int, float, np.float64, np.int64, np.int32]
 Int = Union[int, np.int64, np.int32]
+NDArrayOrScalar = Union[np.ndarray, Sequence[Number], Number]

--- a/dask_ml/_typing.py
+++ b/dask_ml/_typing.py
@@ -1,4 +1,4 @@
-from typing import TypeVar, Union, Sequence
+from typing import Any, Sequence, Type, TypeVar, Union
 
 import dask.dataframe as dd
 import numpy as np
@@ -8,7 +8,7 @@ from pandas import DataFrame, Index, Series
 try:
     from numpy.typing import DTypeLike
 except ImportError:
-    DTypeLike = np.dtype
+    DTypeLike = Any
 
 
 AnyArrayLike = TypeVar("AnyArrayLike", Index, Series, Array, np.ndarray)

--- a/dask_ml/_typing.py
+++ b/dask_ml/_typing.py
@@ -1,10 +1,14 @@
-from typing import TypeVar, Union, Sequence
+from typing import TypeVar, Union, Sequence, Any
 
 import dask.dataframe as dd
 import numpy as np
 from dask.array import Array
 from pandas import DataFrame, Index, Series
-from numpy.typing import DTypeLike
+
+try:
+    from numpy.typing import DTypeLike
+except ImportError:
+    DTypeLike = np.dtype
 
 
 AnyArrayLike = TypeVar("AnyArrayLike", Index, Series, Array, np.ndarray)

--- a/dask_ml/_typing.py
+++ b/dask_ml/_typing.py
@@ -1,4 +1,4 @@
-from typing import TypeVar, Union, Sequence, Any
+from typing import TypeVar, Union, Sequence
 
 import dask.dataframe as dd
 import numpy as np

--- a/dask_ml/metrics/pairwise.py
+++ b/dask_ml/metrics/pairwise.py
@@ -16,7 +16,7 @@ from ..utils import row_norms
 
 
 def pairwise_distances_argmin_min(
-    X: ArrayLike,
+    X: da.Array,
     Y: ArrayLike,
     axis: int = 1,
     metric: Union[str, Callable[[ArrayLike, ArrayLike], float]] = "euclidean",
@@ -47,7 +47,7 @@ def pairwise_distances_argmin_min(
 
 
 def pairwise_distances(
-    X: ArrayLike,
+    X: da.Array,
     Y: ArrayLike,
     metric: Union[str, Callable[[ArrayLike, ArrayLike], float]] = "euclidean",
     n_jobs: Optional[int] = None,
@@ -105,7 +105,7 @@ def euclidean_distances(
 
 
 def check_pairwise_arrays(
-    X: ArrayLike, Y: ArrayLike, precomputed: bool = False
+    X: ArrayLike, Y: Optional[ArrayLike], precomputed: bool = False
 ) -> Tuple[ArrayLike, ArrayLike]:
     # XXX
     if Y is None:

--- a/dask_ml/metrics/regression.py
+++ b/dask_ml/metrics/regression.py
@@ -99,7 +99,8 @@ def r2_score(
     nonzero_denominator = denominator != 0
     nonzero_numerator = numerator != 0
     valid_score = nonzero_denominator & nonzero_numerator
-    output_scores = da.ones([y_true.shape[1]], chunks=y_true.chunks[1])
+    output_chunks = getattr(y_true, "chunks", [None, None])[1]
+    output_scores = da.ones([y_true.shape[1]], chunks=output_chunks)
     with np.errstate(all="ignore"):
         output_scores[valid_score] = 1 - (
             numerator[valid_score] / denominator[valid_score]

--- a/dask_ml/preprocessing/_block_transformer.py
+++ b/dask_ml/preprocessing/_block_transformer.py
@@ -71,7 +71,7 @@ class BlockTransformer(BaseEstimator, TransformerMixin):
         validate: bool = False,
         **kw_args: Any
     ):
-        self.func = func
+        self.func: Callable[..., Union[ArrayLike, DataFrameType]] = func
         self.validate = validate
         self.kw_args = kw_args
 

--- a/dask_ml/preprocessing/_encoders.py
+++ b/dask_ml/preprocessing/_encoders.py
@@ -7,7 +7,7 @@ import pandas as pd
 import sklearn.preprocessing
 
 from .._compat import SK_024
-from .._typing import ArrayLike, DataFrameType, SeriesType, DTypeLike
+from .._typing import ArrayLike, DataFrameType, DTypeLike, SeriesType
 from ..utils import check_array
 from .label import _encode, _encode_dask_array
 

--- a/dask_ml/preprocessing/_encoders.py
+++ b/dask_ml/preprocessing/_encoders.py
@@ -7,7 +7,7 @@ import pandas as pd
 import sklearn.preprocessing
 
 from .._compat import SK_024
-from .._typing import ArrayLike, DataFrameType, SeriesType
+from .._typing import ArrayLike, DataFrameType, SeriesType, DTypeLike
 from ..utils import check_array
 from .label import _encode, _encode_dask_array
 
@@ -116,7 +116,7 @@ class OneHotEncoder(sklearn.preprocessing.OneHotEncoder):
         categories: Union[str, ArrayLike] = "auto",
         drop: Optional[bool] = None,
         sparse: bool = True,
-        dtype: np.dtype = np.float64,
+        dtype: DTypeLike = np.float64,
         handle_unknown: str = "error",
     ):
         if drop is not None:

--- a/dask_ml/preprocessing/data.py
+++ b/dask_ml/preprocessing/data.py
@@ -21,7 +21,7 @@ from dask_ml._compat import blockwise, check_is_fitted
 from dask_ml._utils import copy_learned_attributes
 from dask_ml.utils import check_array, handle_zeros_in_scale
 
-from .._typing import ArrayLike, DataFrameType, SeriesType, NDArrayOrScalar
+from .._typing import ArrayLike, DataFrameType, NDArrayOrScalar, SeriesType
 
 _PANDAS_VERSION = LooseVersion(pd.__version__)
 _HAS_CTD = _PANDAS_VERSION >= "0.21.0"

--- a/dask_ml/preprocessing/data.py
+++ b/dask_ml/preprocessing/data.py
@@ -770,7 +770,7 @@ class DummyEncoder(BaseEstimator, TransformerMixin):
         big = isinstance(X, dd.DataFrame)
 
         if big:
-            divisions = np.ndarray(X.divisions)
+            divisions = np.array(X.divisions)
             divisions[-1] = divisions[-1] + 1
             chunks = tuple(divisions[1:] - divisions[:-1])
 

--- a/dask_ml/preprocessing/data.py
+++ b/dask_ml/preprocessing/data.py
@@ -21,14 +21,14 @@ from dask_ml._compat import blockwise, check_is_fitted
 from dask_ml._utils import copy_learned_attributes
 from dask_ml.utils import check_array, handle_zeros_in_scale
 
-from .._typing import ArrayLike, DataFrameType, SeriesType
+from .._typing import ArrayLike, DataFrameType, SeriesType, NDArrayOrScalar
 
 _PANDAS_VERSION = LooseVersion(pd.__version__)
 _HAS_CTD = _PANDAS_VERSION >= "0.21.0"
 BOUNDS_THRESHOLD = 1e-7
 
 
-def _handle_zeros_in_scale(scale: np.ndarray, copy=True):
+def _handle_zeros_in_scale(scale: NDArrayOrScalar, copy=True):
     """Makes sure that whenever scale is zero, we handle it correctly.
 
     This happens in most scalers when we have constant features."""
@@ -75,7 +75,7 @@ class StandardScaler(sklearn.preprocessing.StandardScaler):
         values = compute(*attributes.values())
         for k, v in zip(attributes, values):
             setattr(self, k, v)
-        self.n_features_in_ = X.shape[1]
+        self.n_features_in_: int = X.shape[1]
         return self
 
     def partial_fit(
@@ -142,7 +142,7 @@ class MinMaxScaler(sklearn.preprocessing.MinMaxScaler):
         values = compute(*attributes.values())
         for k, v in zip(attributes, values):
             setattr(self, k, v)
-        self.n_features_in_ = X.shape[1]
+        self.n_features_in_: int = X.shape[1]
         return self
 
     def partial_fit(
@@ -227,7 +227,7 @@ class RobustScaler(sklearn.preprocessing.RobustScaler):
         self.center_: List[float] = quantiles[:, 1]
         self.scale_: List[float] = quantiles[:, 2] - quantiles[:, 0]
         self.scale_ = _handle_zeros_in_scale(self.scale_, copy=False)
-        self.n_features_in_ = X.shape[1]
+        self.n_features_in_: int = X.shape[1]
         return self
 
     def transform(
@@ -338,6 +338,7 @@ class QuantileTransformer(sklearn.preprocessing.QuantileTransformer):
         references = self.references_ * 100
         quantiles = [da.percentile(col, references) for col in X.T]
         (self.quantiles_,) = compute(da.vstack(quantiles).T)
+        return None
 
     def _transform(
         self, X: Union[ArrayLike, DataFrameType], inverse: bool = False
@@ -352,7 +353,7 @@ class QuantileTransformer(sklearn.preprocessing.QuantileTransformer):
         return da.vstack(transformed, allow_unknown_chunksizes=True).T
 
     def _transform_col(
-        self, X_col: ArrayLike, quantiles: ArrayLike, inverse: bool
+        self, X_col: da.Array, quantiles: ArrayLike, inverse: bool
     ) -> ArrayLike:
         output_distribution = self.output_distribution
 
@@ -761,7 +762,7 @@ class DummyEncoder(BaseEstimator, TransformerMixin):
             if unknown:
                 lengths = blockwise(len, "i", X[:, 0], "i", dtype="i8").compute()
                 X = X.copy()
-                chunks: ArrayLike = (tuple(lengths), X.chunks[1])
+                chunks: tuple = (tuple(lengths), X.chunks[1])
                 X._chunks = chunks
 
             X = dd.from_dask_array(X, columns=self.transformed_columns_)
@@ -769,9 +770,9 @@ class DummyEncoder(BaseEstimator, TransformerMixin):
         big = isinstance(X, dd.DataFrame)
 
         if big:
-            chunks = np.array(X.divisions)
-            chunks[-1] = chunks[-1] + 1
-            chunks = tuple(chunks[1:] - chunks[:-1])
+            divisions = np.ndarray(X.divisions)
+            divisions[-1] = divisions[-1] + 1
+            chunks = tuple(divisions[1:] - divisions[:-1])
 
         non_cat = X[list(self.non_categorical_columns_)]
 
@@ -986,7 +987,7 @@ class OrdinalEncoder(BaseEstimator, TransformerMixin):
             if unknown:
                 lengths = blockwise(len, "i", X[:, 0], "i", dtype="i8").compute()
                 X = X.copy()
-                chunks: ArrayLike = (tuple(lengths), X.chunks[1])
+                chunks: tuple = (tuple(lengths), X.chunks[1])
                 X._chunks = chunks
 
             X = dd.from_dask_array(X, columns=self.columns_)
@@ -994,9 +995,9 @@ class OrdinalEncoder(BaseEstimator, TransformerMixin):
         big = isinstance(X, dd.DataFrame)
 
         if big:
-            chunks = np.array(X.divisions)
-            chunks[-1] = chunks[-1] + 1
-            chunks = tuple(chunks[1:] - chunks[:-1])
+            divisions = np.array(X.divisions)
+            divisions[-1] = divisions[-1] + 1
+            chunks = tuple(divisions[1:] - divisions[:-1])
 
         X = X.copy()
         for col in self.categorical_columns_:

--- a/dask_ml/preprocessing/label.py
+++ b/dask_ml/preprocessing/label.py
@@ -197,7 +197,7 @@ class LabelEncoder(sklearn.preprocessing.LabelEncoder):
 
 
 def _encode_categorical(
-    values: np.ndarray, uniques: np.ndarray = None, encode: bool = False
+    values: pd.Series, uniques: Optional[np.ndarray] = None, encode: bool = False
 ):
     new_uniques = np.asarray(values.cat.categories)
 

